src/protocol/block.rs

```diff
@@ -0,0 +1,216 @@
+#[repr(C)]
+pub struct BlockIOProtocol {
+    revision: u64,
+    pub media: *const BlockIOMedia,
+    reset: unsafe extern "win64" fn(this: *const BlockIOProtocol, extended_verification: u8)
+        -> Status,
+    read_blocks: unsafe extern "win64" fn(
+        this: *const BlockIOProtocol,
+        media_id: u32,
+        lba: u64,
+        buffer_size: usize,
+        buffer: *mut CVoid,
+    ) -> Status,
+    write_blocks: unsafe extern "win64" fn(
+        this: *const BlockIOProtocol,
+        media_id: u32,
+        lba: u64,
+        buffer_size: usize,
+        buffer: *const CVoid,
+    ) -> Status,
+    flush_blocks: unsafe extern "win64" fn(this: *const BlockIOProtocol) -> Status,
+}
+
+impl Protocol for BlockIOProtocol {
+    fn guid() -> &'static Guid {
+        &EFI_BLOCK_IO_PROTOCOL_GUID
+    }
+}
+
+impl BlockIOProtocol {
+    /// Indicates whether or not the device is removable.
+    pub fn is_removable(&self) -> bool {
+        unsafe { (*self.media).removable == 1 }
+    }
+
+    /// Indicates whether or not the device is present.
+    pub fn is_present(&self) -> bool {
+        unsafe { (*self.media).present == 1 }
+    }
+
+    /// Indicates whether or not the device is a logical partition.
+    pub fn is_logical_partition(&self) -> bool {
+        unsafe { (*self.media).logical_partition == 1 }
+    }
+
+    /// Indicates whether or not the device is read only.
+    pub fn is_read_only(&self) -> bool {
+        unsafe { (*self.media).read_only == 1 }
+    }
+
+    /// Indicates whether or not the device performs write caching.
+    pub fn write_caching(&self) -> bool {
+        unsafe { (*self.media).write_caching == 1 }
+    }
+
+    /// Indicates whether or not the device has alignment requirements for buffers.
+    pub fn must_align(&self) -> bool {
+        unsafe { (*self.media).io_align > 1 }
+    }
+
+    pub fn required_pages_block(&self, blocks: usize) -> usize {
+        let block_size = unsafe { (*self.media).block_size } as usize;
+        let mut bytes = block_size * blocks;
+
+        if bytes % block_size != 0 {
+            bytes = block_size * ((bytes / block_size) + 1);
+        }
+
+        self.required_pages(bytes)
+    }
+    pub fn required_pages(&self, read_size: usize) -> usize {
+        let mut num_pages = read_size / 4096;
+        if read_size % 4096 != 0 {
+            num_pages += 1;
+        }
+
+        num_pages
+    }
+
+    /// Reset the device.
+    pub fn reset(&self, extended_verification: bool) -> Result<(), Status> {
+        match unsafe { (self.reset)(self, extended_verification as u8) } {
+            Status::Success => Ok(()),
+            e => Err(e),
+        }
+    }
+
+    /// Read `num_bytes` bytes from the disk starting at block `start`. The returned slice includes
+    /// memory allocated with `allocate_pool` or `allocate_pages`, and it is the caller's
+    /// responsibility to free it with the appropriate function based on `must_align()`.
```

Comment:
This does not make a great API.  Is `allocate_pool` an important optimization?  It seems like it'd be cleaner to always use `allocate_pages`.